src/ir/headers/memories.hpp

```diff
@@ -282,20 +283,63 @@ Namespace* CoreIRLoadHeader_memory(Context* c) {
-    def->addInstance("waddr0","coreir.const",{{"width",Const::make(c,awidth)}},{{"value",Const::make(c,0)}});
+    def->addInstance("readreg","mantle.reg",{{"width",Const::make(c,width)},{"has_en",Const::make(c,true)}});
+    def->addInstance("wdata0","coreir.const",{{"width",Const::make(c,width)}},{{"value",Const::make(c,BitVector(width,0))}});
+    def->addInstance("waddr0","coreir.const",{{"width",Const::make(c,awidth)}},{{"value",Const::make(c,BitVector(awidth,0))}});
     def->connect("self.clk","mem.clk");
     def->connect("self.clk","readreg.clk");
-    def->connect("wdata0","mem.wdata");
-    def->connect("waddr0","mem.waddr");
-    def->connect("wdata0.0","mem.wen");
+    def->connect("wdata0.out","mem.wdata");
+    def->connect("waddr0.out","mem.waddr");
+    def->connect("wdata0.out.0","mem.wen");
     def->connect("mem.rdata","readreg.in");
     def->connect("self.rdata","readreg.out");
     def->connect("self.raddr","mem.raddr");
     def->connect("self.ren","readreg.en");
   });
 
+  memory->newTypeGen("Rom2Type",MemGenParams,[](Context* c, Values genargs) {
+    uint width = genargs.at("width")->get<int>();
+    return c->Record({
+      {"clk", c->Named("coreir.clkIn")},
+      {"rdata", c->Bit()->Arr(width)},
+      {"raddr", c->BitIn()->Arr(width)},
+      {"ren", c->BitIn()},
+    });
+  });
+
+
+  Generator* rom2 = memory->newGeneratorDecl("rom2",memory->getTypeGen("Rom2Type"),MemGenParams);
+  rom2->setModParamsGen(RomModParamFun);
+  rom2->setGeneratorDefFromFun([](Context* c, Values genargs, ModuleDef* def) {
+    uint width = genargs.at("width")->get<int>();
+    uint depth = genargs.at("depth")->get<int>();
+    uint awidth = (uint) ceil(log2(depth));
+    
+    Values memargs = genargs;
+    memargs.insert({"has_init",Const::make(c,true)});
+
+    def->addInstance("mem","coreir.mem",memargs,{{"init",def->getModule()->getArg("init")}});
+    def->addInstance("readreg","mantle.reg",{{"width",Const::make(c,width)},{"has_en",Const::make(c,true)}});
+    def->addInstance("wdata0","coreir.const",{{"width",Const::make(c,width)}},{{"value",Const::make(c,BitVector(width,0))}});
+    def->addInstance("waddr0","coreir.const",{{"width",Const::make(c,awidth)}},{{"value",Const::make(c,BitVector(awidth,0))}});
+    Values sliceArgs = {{"width", Const::make(c,width)},
+                        {"lo", Const::make(c,0)},
+                        {"hi", Const::make(c,awidth)}};
+    def->addInstance("raddr_slice","coreir.slice",sliceArgs);
+                     
+    def->connect("self.clk","mem.clk");
+    def->connect("self.clk","readreg.clk");
+    def->connect("wdata0.out","mem.wdata");
+    def->connect("waddr0.out","mem.waddr");
+    def->connect("wdata0.out.0","mem.wen");
+    def->connect("mem.rdata","readreg.in");
+    def->connect("self.rdata","readreg.out");
+    def->connect("self.raddr","raddr_slice.in");
+    def->connect("raddr_slice.out","mem.raddr");
+    def->connect("self.ren","readreg.en");
+  });
+
+  
```

Comment:
Adding this new generator shouldn't affect any existing code

Reply:
Can you add a comment explaining how this rom (rom2) is different than the normal rom? Also, is there a better (more descriptive) name than just `rom2`?

Reply:
From what I understand, this module sets waddr/wdata/wen to 0 (since it's a rom), then passes the rdata through a register (adds a single cycle delay). It also seems to slice the `raddr` (so it only selects the lower bits based on the depth). Why can't the interface to this ROM have the `raddr` width equal to clog2(depth)? This would remove the need for the slice module.

Reply:
Maybe `rom_delay_1` as a name (we should figure out a better way to handle parametrized delays, but this is okay for now to unblock you).

Reply:
Code was modified to the existing generators, because after usage, they were not correct; these changes allow them to compile correctly. rom2 uses a port that is the same bitwdith as the data (even though this is more bits than necessary).